libs/cardano-ledger-core/src/Cardano/Ledger/Core/Era.hs

```diff
@@ -106,6 +114,37 @@ instance Crypto c => Era (ByronEra c) where
 
   eraName = "Byron"
 
+-----------------------------
+-- Rules --------------------
+-----------------------------
+
+-- | Era STS map
+type family EraRule (rule :: Symbol) era = (r :: Type) | r -> rule
+
+-- | `EraRuleFailure` type family is needed for injectivity, which STS does not provide for
+-- us unfortunately.
+type family EraRuleFailure (rule :: Symbol) era = (r :: Type) | r -> rule era
+
+data VoidRuleFailure (rule :: Symbol) era
+
+-- Rules that should never have a predicate failures
+type instance EraRuleFailure "EPOCH" era = VoidRuleFailure "EPOCH" era
+type instance EraRuleFailure "NEWEPOCH" era = VoidRuleFailure "NEWEPOCH" era
+type instance EraRuleFailure "MIR" era = VoidRuleFailure "MIR" era
+type instance EraRuleFailure "NEWPP" era = VoidRuleFailure "NEWPP" era
+type instance EraRuleFailure "SNAP" era = VoidRuleFailure "SNAP" era
+type instance EraRuleFailure "TICK" era = VoidRuleFailure "TICK" era
+type instance EraRuleFailure "TICKF" era = VoidRuleFailure "TICKF" era
+type instance EraRuleFailure "UPEC" era = VoidRuleFailure "UPEC" era
+type instance EraRuleFailure "RUPD" era = VoidRuleFailure "RUPD" era
+type instance EraRuleFailure "POOLREAP" era = VoidRuleFailure "POOLREAP" era
+
+class
+  EraRuleFailure rule era ~ PredicateFailure (EraRule rule era) =>
+  InjectRuleFailure (rule :: Symbol) t era
+  where
+  injectFailure :: t era -> EraRuleFailure rule era
```

Comment:
```suggestion
  injectFailure :: t era -> EraRuleFailure rule era
  default injectFailure :: t era ~ EraRuleFailure rule era => t era -> EraRuleFailure rule era
  injectFailure = id
```

This might get rid of some of the boilerplate

Reply:
Yeah, this is true. Thanks